### ISSUE TITLE
Weird stuttering and generally low quality when using with Cloudflare Calls

### ISSUE THREAD

**Issue by skirsten** [react-native-webrtc/react-native-webrtc#1536]
Hey, just today Cloudflare released their Calls product into [open beta](https://blog.cloudflare.com/cloudflare-calls-anycast-webrtc) so I thought I would check it out on Expo with this awesome library.

I ported their example as closely as I could to this library. And unfortunately there is some weirdness going on.

Further down I attached a video where you can see that the stream is frozen every ~3 seconds or so. This does not happen in their html example running in Chrome on the same phone. Also the quality is much worse than in their example.

I attached the [cf-calls-demo.html](https://github.com/react-native-webrtc/react-native-webrtc/files/14877684/cf-calls-demo.html.txt) that you get on the CF dashboard after creating a new Calls app.
Also here is the relevant ported code:

```typescript
import React, { useState } from "react";
import { Button, StyleSheet, Text, View } from "react-native";
import {
  MediaStream,
  MediaStreamTrack,
  RTCPeerConnection,
  RTCRtpTransceiver,
  RTCSessionDescription,
  RTCView,
  mediaDevices,
} from "react-native-webrtc";
import { assert } from "tsafe/assert";
import { CallsApp } from "./utils/CallsApp";

class WebRTCApp {
  pc: RTCPeerConnection;
  app: CallsApp;
  transceivers: RTCRtpTransceiver[];

  constructor(appId: string) {
    // Use Cloudflare's STUN server
    this.pc = new RTCPeerConnection({
      iceServers: [
        {
          urls: "stun:stun.cloudflare.com:3478",
        },
      ],
      bundlePolicy: "max-bundle",
    });

    // Create a instance of CallsApp (defined below). Please note that this is not an official SDK but just a demo showing the HTML API.
    this.app = new CallsApp(appId);
  }

  async setup() {
    // In order to successfully establish a peer connection, we need at least one track to publish.
    // In this case, we create two: video & audio
    const localStream = await mediaDevices.getUserMedia({
      video: {
        // frameRate: 30,
        facingMode: "environment",
      },
      audio: true,
    });

    // return localStream;

    // Add sendonly trancievers to the PeerConnection
    this.transceivers = localStream.getTracks().map((track) =>
      this.pc.addTransceiver(track, {
        direction: "sendonly",
      })
    );

    const sessionConstraints = undefined;
    // TODO: Why is this included in the example?
    // {
    //   mandatory: {
    //     OfferToReceiveAudio: true,
    //     OfferToReceiveVideo: true,
    //     VoiceActivityDetection: true,
    //   },
    // };

    // Send the first offer and create a session. The returned sessionId is required to retrieve any track published by this peer
    await this.pc.setLocalDescription(
      await this.pc.createOffer(sessionConstraints)
    );

    assert(this.pc.localDescription);

    const newSessionResult = await this.app.newSession(
      this.pc.localDescription.sdp
    );

    await this.pc.setRemoteDescription(
      new RTCSessionDescription(newSessionResult.sessionDescription)
    );

    // Make sure the peer connection was established
    await new Promise((resolve, reject) => {
      this.pc.addEventListener("iceconnectionstatechange", (ev) => {
        if (this.pc.iceConnectionState === "connected") {
          resolve(undefined);
        }
        setTimeout(reject, 5000, "connect timeout");
      });
    });

    // We associate a trackName to a transceiver identified by a mid (media ID). This way the track
    // is remotely reachable by the tuple (sessionId, trackName)
    let trackObjects: any[] = this.transceivers.map((transceiver) => {
      assert(transceiver.sender.track);

      return {
        location: "local",
        mid: transceiver.mid,
        trackName: transceiver.sender.track.id,
      };
    });

    // Get local description, create a new track, set remote description with the response
    await this.pc.setLocalDescription(
      await this.pc.createOffer(sessionConstraints)
    );

    assert(this.pc.localDescription);

    const newLocalTracksResult = await this.app.newTracks(
      trackObjects,
      this.pc.localDescription.sdp
    );

    await this.pc.setRemoteDescription(
      new RTCSessionDescription(newLocalTracksResult.sessionDescription)
    );

    // At this point in code, we are successfully sending local stream to Cloudflare Calls.
    // Now, we will pull the same stream from Cloudflare Calls.

    // Update trackObjects to reference the tracks as "remote"
    trackObjects = trackObjects.map((trackObject) => {
      return {
        location: "remote",
        sessionId: this.app.sessionId,
        trackName: trackObject.trackName,
      };
    });

    // Prepare to receive the tracks before asking for them
    const remoteTracksPromise = new Promise<MediaStreamTrack[]>((resolve) => {
      let tracks: MediaStreamTrack[] = [];

      this.pc.addEventListener("track", (event) => {
        assert(event.track);

        tracks.push(event.track);
        console.debug(`Got track mid=${event.track["mid"]}`);

        if (tracks.length >= 2) {
          // remote video & audio are ready
          resolve(tracks);
        }
      });
    });

    // Calls API request to ask for the tracks
    const newRemoteTracksResult = await this.app.newTracks(trackObjects);
    if (newRemoteTracksResult.requiresImmediateRenegotiation) {
      switch (newRemoteTracksResult.sessionDescription.type) {
        case "offer":
          // We let Cloudflare know we're ready to receive the tracks
          await this.pc.setRemoteDescription(
            new RTCSessionDescription(newRemoteTracksResult.sessionDescription)
          );
          await this.pc.setLocalDescription(await this.pc.createAnswer());
          await this.app.sendAnswerSDP(this.pc.localDescription.sdp);
          break;
        case "answer":
          throw new Error("An offer SDP was expected");
      }
    }

    // Once started receiving the tracks (video & audio) send the data to the video tag
    const remoteTracks = await remoteTracksPromise;

    const remoteStream = new MediaStream();
    remoteStream.addTrack(remoteTracks[0]);
    remoteStream.addTrack(remoteTracks[1]);

    return remoteStream;
  }

  async shutdown() {
    // TODO:
    // localMediaStream.getTracks().forEach(
    //   track => track.stop()
    // );
    // this.pc.close();
  }
}

const app = new WebRTCApp("xxx");

export default function Page() {
  const [stream, setStream] = useState(null);

  const start = async () => {
    console.log("start");
    if (!stream) {
      try {
        const s = await app.setup();

        setStream(s);
      } catch (e) {
        console.error(e);
      }
    }
  };

  const stop = () => {
    console.log("stop");
    if (stream) {
      stream.release();
      setStream(null);
    }
  };

  return (
    <View style={styles.container}>
      <View style={styles.main}>
        <Text style={styles.title}>Hello World</Text>
        {/* <Text style={styles.subtitle}>This is the first page of your app.</Text> */}

        {stream && <RTCView streamURL={stream.toURL()} style={styles.stream} />}

        <Button title="Start" onPress={start} />
        <Button title="Stop" onPress={stop} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    padding: 24,
  },
  main: {
    flex: 1,
    justifyContent: "center",
    maxWidth: 960,
    marginHorizontal: "auto",
  },
  title: {
    fontSize: 64,
    fontWeight: "bold",
  },
  subtitle: {
    fontSize: 36,
    color: "#38434D",
  },
  stream: {
    flex: 1,
  },
});
```

Also here is a video that should make the quality and stuttering obvious:

https://github.com/react-native-webrtc/react-native-webrtc/assets/1972314/0f7225ff-511b-4613-986e-0966d0de0058

Also here are all the logs:

```log

 LOG  rn-webrtc:pc:DEBUG 0 ctor +0ms
 LOG  start
 LOG  rn-webrtc:pc:DEBUG 0 addTransceiver +2s
 LOG  rn-webrtc:pc:DEBUG 0 addTransceiver +21ms
 LOG  rn-webrtc:pc:DEBUG 0 createOffer +10ms
 LOG  rn-webrtc:pc:DEBUG 0 createOffer OK +9ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription +2ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription OK +37ms
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription +879ms
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription OK +13ms
 LOG  rn-webrtc:pc:DEBUG 0 createOffer +764ms
 LOG  rn-webrtc:pc:DEBUG 0 createOffer OK +100ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription +0ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription OK +17ms
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription +356ms
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription OK +22ms
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription +153ms
 LOG  rn-webrtc:pc:DEBUG 0 ontrack +29ms
 LOG  rn-webrtc:pc:DEBUG 0 ontrack +6ms
 DEBUG  Got track mid=undefined
 DEBUG  Got track mid=undefined
 LOG  rn-webrtc:pc:DEBUG 0 setRemoteDescription OK +14ms
 LOG  rn-webrtc:pc:DEBUG 0 createAnswer +1ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription +13ms
 LOG  rn-webrtc:pc:DEBUG 0 setLocalDescription OK +142ms

```

I hope you have some ideas what could be going wrong here or what I need to implement / improve. Thanks :smiley: 

#### Platform Information  
* **React Native Version**: 0.73.6
* **WebRTC Module Version**: 118.0.4
* **Platform OS + Version**: Expo on Android 50.0.14

**Comment by 8BallBomBom** [comment]
Interesting 🤔 
Been waiting to see when they would finally release a beta for their WebRTC services.
I haven't had a decent read as of yet but the fact you've managed to get anything connected is reasonable.
Their SFU should be decent for use on a global scale.

As for the quality issue, that can be due to several reasons.
It is recommended to use a TURN server to relay call data and get a decent connection between peers.
That being said as you are likely doing a peer connection locally to external and then external back to local it would indicate your connection to and from Cloudflare and back might not be too good.
But then again their SFU might also be overwhelmed or even limited as it has only just been opened up for general use.

**Comment by saghul** [comment]
Thanks for the heads up!

I don't see anything obvious right off the bat.

Something interesting to check is what codec gets negotiated.

**Comment by skirsten** [comment]
Here is the report of the "inbound-rtp" "video":

```json
{
  "report": {
    "timestamp": 1712341207066.357,
    "type": "inbound-rtp",
    "id": "IT01V90996005",
    "ssrc": 90996005,
    "kind": "video",
    "transportId": "T01",
    "codecId": "CIT01_98_level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f",
    "jitter": 0.11,
    "packetsLost": 72,
    "trackIdentifier": "XJfhUjGqQflicKzW",
    "mid": "1",
    "packetsReceived": 2106,
    "bytesReceived": 1830148,
    "headerBytesReceived": 42120,
    "lastPacketReceivedTimestamp": 1712341205934.252,
    "jitterBufferDelay": 405.635804,
    "jitterBufferTargetDelay": 61.455742,
    "jitterBufferMinimumDelay": 61.455742,
    "jitterBufferEmittedCount": 704,
    "framesReceived": 520,
    "frameWidth": 176,
    "frameHeight": 320,
    "framesDecoded": 520,
    "keyFramesDecoded": 10,
    "framesDropped": 300,
    "totalDecodeTime": 2.546,
    "totalProcessingDelay": 378.347062,
    "totalAssemblyTime": 8.125407,
    "framesAssembledFromMultiplePackets": 218,
    "totalInterFrameDelay": 44.744,
    "totalSquaredInterFrameDelay": 68.84777000000008,
    "pauseCount": 0,
    "totalPausesDuration": 0,
    "freezeCount": 14,
    "totalFreezesDuration": 23.094,
    "decoderImplementation": "OMX.qcom.video.decoder.avc",
    "firCount": 0,
    "pliCount": 4,
    "nackCount": 20,
    "qpSum": 18449,
    "powerEfficientDecoder": true,
    "minPlayoutDelay": 0
  },
  "codec": {
    "timestamp": 1712341207066.357,
    "type": "codec",
    "id": "CIT01_98_level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f",
    "transportId": "T01",
    "payloadType": 98,
    "mimeType": "video/H264",
    "clockRate": 90000,
    "sdpFmtpLine": "level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f"
  }
}
```

Its using H264 as codec. Its actually starting in a better resolution but after some time it just gets worse and worse (here it landed at 176x320 resolution).

In comparison the same code running in Google Chrome on the same phone with the same network speed etc. is completely stable and using VP8. After minutes of use the resolution stays stable and the errors are not increasing:
```
video/VP8, freezeCount=2, nackCount=0, framesDropped=1
```

Maybe this library is handling some things like NACK's differently causing it to think it needs to constantly decrease quality?

What can be excluded:
- TURN: No need here as the SFU is anycast globally routable IP's and I don't have a strict egress firewall
- Connection to CF: Considering that I am running the exact same test on the same phone with the same connectivity etc. this can also be excluded.
- overwhelmed SFU: :point_up: Same as above, the SFU and connection to it is working great on many devices using Chrome (including the same phone).

I was hoping that it was some kind of simple error on my side (or something on React re-rendering causing the issues) but unfortunately I triple checked all that and I don't think its any of that.

**Comment by saghul** [comment]
Can you try to use VP8 instead? That would make it an apples to apples comparison.

**Comment by 8BallBomBom** [comment]
@skirsten Thanks for doing some research and testing, always best to cover as many bases as possible 😄
Can't say much about the inner working of NACK but you are correct, if calculated wrong it can be assumed connection quality is bad which would result in the quality being downgraded.
Might be worth trying to compare.

But as mentioned above ^ can you run some more tests with forced usage of VP8.
Could just be the H264 hardware codec being faulty, seems to be a common issue with plenty of devices.

**Comment by renandincer** [comment]
Hey! I work on the Calls team. Happy to help on this if you need info from the SFU side.

**Comment by saghul** [comment]
Thanks for dropping by! We'll certainly reach out!

**Comment by skirsten** [comment]
Hey,

I forced the codec to VP8 and the issue persists. It starts off with pretty okay quality but then every ~3 seconds or so it degrades the quality (which is also what is causing the freezes probably). On chrome its the other way around, it starts with pretty okay quality and gradually improves it until its great.

<details><summary>Here the info dump from this library</summary>

```json
[
  {
    "report": {
      "id": "IT01V2437779428",
      "timestamp": 1712416967279.634,
      "type": "inbound-rtp",
      "codecId": "CIT01_100",
      "kind": "video",
      "ssrc": 2437779428,
      "transportId": "T01",
      "jitter": 0.014,
      "packetsLost": 0,
      "packetsReceived": 2050,
      "bytesReceived": 1688458,
      "decoderImplementation": "libvpx",
      "firCount": 0,
      "frameHeight": 320,
      "frameWidth": 180,
      "framesAssembledFromMultiplePackets": 595,
      "framesDecoded": 1165,
      "framesDropped": 10,
      "framesPerSecond": 31,
      "framesReceived": 1165,
      "freezeCount": 9,
      "headerBytesReceived": 41028,
      "jitterBufferDelay": 232.59160899999998,
      "jitterBufferEmittedCount": 1165,
      "jitterBufferMinimumDelay": 134.12340799999998,
      "jitterBufferTargetDelay": 134.12340799999998,
      "keyFramesDecoded": 12,
      "lastPacketReceivedTimestamp": 1712416967277.774,
      "mid": "1",
      "nackCount": 2,
      "pauseCount": 0,
      "pliCount": 11,
      "powerEfficientDecoder": false,
      "totalAssemblyTime": 9.656922999999999,
      "totalDecodeTime": 1.201341,
      "totalFreezesDuration": 10.807,
      "totalInterFrameDelay": 41.616,
      "totalPausesDuration": 0,
      "totalProcessingDelay": 234.15373799999998,
      "totalSquaredInterFrameDelay": 15.845633999999977,
      "trackIdentifier": "rMyoBBlLrbiGgPCU",
      "minPlayoutDelay": 0,
      "qpSum": 84979
    },
    "codec": {
      "id": "CIT01_100",
      "timestamp": 1712416967279.634,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 100,
      "transportId": "T01"
    }
  },
  {
    "report": {
      "id": "OT01V1389257629",
      "timestamp": 1712416967279.634,
      "type": "outbound-rtp",
      "codecId": "COT01_100",
      "kind": "video",
      "ssrc": 1389257629,
      "transportId": "T01",
      "bytesSent": 1707820,
      "packetsSent": 2079,
      "active": true,
      "encoderImplementation": "libvpx",
      "firCount": 0,
      "frameHeight": 320,
      "frameWidth": 180,
      "framesEncoded": 1199,
      "framesPerSecond": 31,
      "framesSent": 1199,
      "headerBytesSent": 46529,
      "hugeFramesSent": 22,
      "keyFramesEncoded": 14,
      "mediaSourceId": "SV1",
      "mid": "0",
      "nackCount": 0,
      "pliCount": 10,
      "powerEfficientEncoder": false,
      "qpSum": 88042,
      "qualityLimitationDurations": {
        "bandwidth": 47.311,
        "cpu": 0,
        "none": 0.044,
        "other": 0
      },
      "qualityLimitationReason": "bandwidth",
      "qualityLimitationResolutionChanges": 0,
      "remoteId": "RIV1389257629",
      "retransmittedBytesSent": 0,
      "retransmittedPacketsSent": 0,
      "targetBitrate": 600000,
      "totalEncodeTime": 5.381,
      "totalEncodedBytesTarget": 0,
      "totalPacketSendDelay": 49.149696999999996,
      "scalabilityMode": "L1T1"
    },
    "codec": {
      "id": "COT01_100",
      "timestamp": 1712416967279.634,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 100,
      "transportId": "T01"
    }
  },
  {
    "report": {
      "id": "RIV1389257629",
      "timestamp": 1712416966599,
      "type": "remote-inbound-rtp",
      "codecId": "COT01_100",
      "kind": "video",
      "ssrc": 1389257629,
      "transportId": "T01",
      "jitter": 0.010744,
      "packetsLost": 0,
      "fractionLost": 0,
      "localId": "OT01V1389257629",
      "roundTripTimeMeasurements": 0,
      "totalRoundTripTime": 0
    },
    "codec": {
      "id": "COT01_100",
      "timestamp": 1712416967279.634,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 100,
      "transportId": "T01"
    }
  }
]
```
</details>

<details><summary>Info dump from Chrome</summary>

```json
[
  {
    "report": {
      "id": "IT01V2611414147",
      "timestamp": 1712417743833.763,
      "type": "inbound-rtp",
      "codecId": "CIT01_96",
      "kind": "video",
      "mediaType": "video",
      "ssrc": 2611414147,
      "transportId": "T01",
      "jitter": 0.015,
      "packetsLost": 0,
      "packetsReceived": 1652,
      "bytesReceived": 1599481,
      "decoderImplementation": "ExternalDecoder (MediaCodecVideoDecoder)",
      "firCount": 0,
      "frameHeight": 480,
      "frameWidth": 270,
      "framesAssembledFromMultiplePackets": 501,
      "framesDecoded": 592,
      "framesDropped": 0,
      "framesPerSecond": 31,
      "framesReceived": 592,
      "freezeCount": 0,
      "headerBytesReceived": 33040,
      "jitterBufferDelay": 25.276087,
      "jitterBufferEmittedCount": 592,
      "jitterBufferMinimumDelay": 27.502305,
      "jitterBufferTargetDelay": 27.502305,
      "keyFramesDecoded": 6,
      "lastPacketReceivedTimestamp": 1712417643660.086,
      "mid": "1",
      "nackCount": 5,
      "pauseCount": 0,
      "pliCount": 2,
      "powerEfficientDecoder": true,
      "totalAssemblyTime": 2.914541,
      "totalDecodeTime": 3.973982,
      "totalFreezesDuration": 0,
      "totalInterFrameDelay": 19.672,
      "totalPausesDuration": 0,
      "totalProcessingDelay": 29.369705,
      "totalSquaredInterFrameDelay": 0.7589100000000008,
      "trackIdentifier": "hPyvkXzbqebvcRPn"
    },
    "codec": {
      "id": "CIT01_96",
      "timestamp": 1712417743833.763,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 96,
      "transportId": "T01"
    }
  },
  {
    "report": {
      "id": "OT01V2135906618",
      "timestamp": 1712417743833.763,
      "type": "outbound-rtp",
      "codecId": "COT01_96",
      "kind": "video",
      "mediaType": "video",
      "ssrc": 2135906618,
      "transportId": "T01",
      "bytesSent": 1651909,
      "packetsSent": 1707,
      "active": true,
      "encoderImplementation": "ExternalEncoder",
      "firCount": 0,
      "frameHeight": 480,
      "frameWidth": 270,
      "framesEncoded": 620,
      "framesPerSecond": 30,
      "framesSent": 620,
      "headerBytesSent": 34591,
      "hugeFramesSent": 1,
      "keyFramesEncoded": 7,
      "mediaSourceId": "SV6",
      "mid": "0",
      "nackCount": 1,
      "pliCount": 5,
      "powerEfficientEncoder": true,
      "qpSum": 42499,
      "qualityLimitationDurations": {
        "bandwidth": 22.13,
        "cpu": 0,
        "none": 0.02,
        "other": 0
      },
      "qualityLimitationReason": "bandwidth",
      "qualityLimitationResolutionChanges": 0,
      "remoteId": "RIV2135906618",
      "retransmittedBytesSent": 1150,
      "retransmittedPacketsSent": 1,
      "rtxSsrc": 1845120888,
      "targetBitrate": 1426216,
      "totalEncodeTime": 4.99,
      "totalEncodedBytesTarget": 0,
      "totalPacketSendDelay": 21.971479
    },
    "codec": {
      "id": "COT01_96",
      "timestamp": 1712417743833.763,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 96,
      "transportId": "T01"
    }
  },
  {
    "report": {
      "id": "RIV2135906618",
      "timestamp": 1712417643135,
      "type": "remote-inbound-rtp",
      "codecId": "COT01_96",
      "kind": "video",
      "mediaType": "video",
      "ssrc": 2135906618,
      "transportId": "T01",
      "jitter": 0.0040999999999999995,
      "packetsLost": 1,
      "fractionLost": 0,
      "localId": "OT01V2135906618",
      "roundTripTimeMeasurements": 0,
      "totalRoundTripTime": 0
    },
    "codec": {
      "id": "COT01_96",
      "timestamp": 1712417743833.763,
      "type": "codec",
      "clockRate": 90000,
      "mimeType": "video/VP8",
      "payloadType": 96,
      "transportId": "T01"
    }
  }
]
```
</details>

Here the diff between the two:
```diff
--- info-chrome.json	2024-04-06 17:40:49.343685391 +0200
+++ info-react-native-ordered.json	2024-04-06 17:40:50.303759658 +0200
@@ -1,140 +1,139 @@
 [
   {
     "report": {
-      "id": "IT01V2611414147",
-      "timestamp": 1712417743833.763,
+      "id": "IT01V2437779428",
+      "timestamp": 1712416967279.634,
       "type": "inbound-rtp",
-      "codecId": "CIT01_96",
+      "codecId": "CIT01_100",
       "kind": "video",
-      "mediaType": "video",
-      "ssrc": 2611414147,
+      "ssrc": 2437779428,
       "transportId": "T01",
-      "jitter": 0.015,
+      "jitter": 0.014,
       "packetsLost": 0,
-      "packetsReceived": 1652,
-      "bytesReceived": 1599481,
-      "decoderImplementation": "ExternalDecoder (MediaCodecVideoDecoder)",
+      "packetsReceived": 2050,
+      "bytesReceived": 1688458,
+      "decoderImplementation": "libvpx",
       "firCount": 0,
-      "frameHeight": 480,
-      "frameWidth": 270,
-      "framesAssembledFromMultiplePackets": 501,
-      "framesDecoded": 592,
-      "framesDropped": 0,
+      "frameHeight": 320,
+      "frameWidth": 180,
+      "framesAssembledFromMultiplePackets": 595,
+      "framesDecoded": 1165,
+      "framesDropped": 10,
       "framesPerSecond": 31,
-      "framesReceived": 592,
-      "freezeCount": 0,
-      "headerBytesReceived": 33040,
-      "jitterBufferDelay": 25.276087,
-      "jitterBufferEmittedCount": 592,
-      "jitterBufferMinimumDelay": 27.502305,
-      "jitterBufferTargetDelay": 27.502305,
-      "keyFramesDecoded": 6,
-      "lastPacketReceivedTimestamp": 1712417643660.086,
+      "framesReceived": 1165,
+      "freezeCount": 9,
+      "headerBytesReceived": 41028,
+      "jitterBufferDelay": 232.59160899999998,
+      "jitterBufferEmittedCount": 1165,
+      "jitterBufferMinimumDelay": 134.12340799999998,
+      "jitterBufferTargetDelay": 134.12340799999998,
+      "keyFramesDecoded": 12,
+      "lastPacketReceivedTimestamp": 1712416967277.774,
       "mid": "1",
-      "nackCount": 5,
+      "nackCount": 2,
       "pauseCount": 0,
-      "pliCount": 2,
-      "powerEfficientDecoder": true,
-      "totalAssemblyTime": 2.914541,
-      "totalDecodeTime": 3.973982,
-      "totalFreezesDuration": 0,
-      "totalInterFrameDelay": 19.672,
+      "pliCount": 11,
+      "powerEfficientDecoder": false,
+      "totalAssemblyTime": 9.656922999999999,
+      "totalDecodeTime": 1.201341,
+      "totalFreezesDuration": 10.807,
+      "totalInterFrameDelay": 41.616,
       "totalPausesDuration": 0,
-      "totalProcessingDelay": 29.369705,
-      "totalSquaredInterFrameDelay": 0.7589100000000008,
-      "trackIdentifier": "hPyvkXzbqebvcRPn"
+      "totalProcessingDelay": 234.15373799999998,
+      "totalSquaredInterFrameDelay": 15.845633999999977,
+      "trackIdentifier": "rMyoBBlLrbiGgPCU",
+      "minPlayoutDelay": 0,
+      "qpSum": 84979
     },
     "codec": {
-      "id": "CIT01_96",
-      "timestamp": 1712417743833.763,
+      "id": "CIT01_100",
+      "timestamp": 1712416967279.634,
       "type": "codec",
       "clockRate": 90000,
       "mimeType": "video/VP8",
-      "payloadType": 96,
+      "payloadType": 100,
       "transportId": "T01"
     }
   },
   {
     "report": {
-      "id": "OT01V2135906618",
-      "timestamp": 1712417743833.763,
+      "id": "OT01V1389257629",
+      "timestamp": 1712416967279.634,
       "type": "outbound-rtp",
-      "codecId": "COT01_96",
+      "codecId": "COT01_100",
       "kind": "video",
-      "mediaType": "video",
-      "ssrc": 2135906618,
+      "ssrc": 1389257629,
       "transportId": "T01",
-      "bytesSent": 1651909,
-      "packetsSent": 1707,
+      "bytesSent": 1707820,
+      "packetsSent": 2079,
       "active": true,
-      "encoderImplementation": "ExternalEncoder",
+      "encoderImplementation": "libvpx",
       "firCount": 0,
-      "frameHeight": 480,
-      "frameWidth": 270,
-      "framesEncoded": 620,
-      "framesPerSecond": 30,
-      "framesSent": 620,
-      "headerBytesSent": 34591,
-      "hugeFramesSent": 1,
-      "keyFramesEncoded": 7,
-      "mediaSourceId": "SV6",
+      "frameHeight": 320,
+      "frameWidth": 180,
+      "framesEncoded": 1199,
+      "framesPerSecond": 31,
+      "framesSent": 1199,
+      "headerBytesSent": 46529,
+      "hugeFramesSent": 22,
+      "keyFramesEncoded": 14,
+      "mediaSourceId": "SV1",
       "mid": "0",
-      "nackCount": 1,
-      "pliCount": 5,
-      "powerEfficientEncoder": true,
-      "qpSum": 42499,
+      "nackCount": 0,
+      "pliCount": 10,
+      "powerEfficientEncoder": false,
+      "qpSum": 88042,
       "qualityLimitationDurations": {
-        "bandwidth": 22.13,
+        "bandwidth": 47.311,
         "cpu": 0,
-        "none": 0.02,
+        "none": 0.044,
         "other": 0
       },
       "qualityLimitationReason": "bandwidth",
       "qualityLimitationResolutionChanges": 0,
-      "remoteId": "RIV2135906618",
-      "retransmittedBytesSent": 1150,
-      "retransmittedPacketsSent": 1,
-      "rtxSsrc": 1845120888,
-      "targetBitrate": 1426216,
-      "totalEncodeTime": 4.99,
+      "remoteId": "RIV1389257629",
+      "retransmittedBytesSent": 0,
+      "retransmittedPacketsSent": 0,
+      "targetBitrate": 600000,
+      "totalEncodeTime": 5.381,
       "totalEncodedBytesTarget": 0,
-      "totalPacketSendDelay": 21.971479
+      "totalPacketSendDelay": 49.149696999999996,
+      "scalabilityMode": "L1T1"
     },
     "codec": {
-      "id": "COT01_96",
-      "timestamp": 1712417743833.763,
+      "id": "COT01_100",
+      "timestamp": 1712416967279.634,
       "type": "codec",
       "clockRate": 90000,
       "mimeType": "video/VP8",
-      "payloadType": 96,
+      "payloadType": 100,
       "transportId": "T01"
     }
   },
   {
     "report": {
-      "id": "RIV2135906618",
-      "timestamp": 1712417643135,
+      "id": "RIV1389257629",
+      "timestamp": 1712416966599,
       "type": "remote-inbound-rtp",
-      "codecId": "COT01_96",
+      "codecId": "COT01_100",
       "kind": "video",
-      "mediaType": "video",
-      "ssrc": 2135906618,
+      "ssrc": 1389257629,
       "transportId": "T01",
-      "jitter": 0.0040999999999999995,
-      "packetsLost": 1,
+      "jitter": 0.010744,
+      "packetsLost": 0,
       "fractionLost": 0,
-      "localId": "OT01V2135906618",
+      "localId": "OT01V1389257629",
       "roundTripTimeMeasurements": 0,
       "totalRoundTripTime": 0
     },
     "codec": {
-      "id": "COT01_96",
-      "timestamp": 1712417743833.763,
+      "id": "COT01_100",
+      "timestamp": 1712416967279.634,
       "type": "codec",
       "clockRate": 90000,
       "mimeType": "video/VP8",
-      "payloadType": 96,
+      "payloadType": 100,
       "transportId": "T01"
     }
   }
 ]
```

I hope this is of some help. I can see some values are much higher (e.g. the jitter stuff), but no idea why.
Also interesting that the traffic (see status bar in the video) goes up with the quality dropping :confused:
[Screenrecorder-2024-04-06-17-08-19-740.webm](https://github.com/react-native-webrtc/react-native-webrtc/assets/1972314/c735202c-a233-41be-94b2-962a391359af)

**Comment by renandincer** [comment]
FWIW, Calls (when it doesn't know better, as a fallback mechanism) will send a PLI regularly at ~5sec intervals - but this shouldn't cause different behavior between the two systems you're comparing.

**Comment by saghul** [comment]
What does it need in order not to do that? We just wrap libwebrtc 118 here. FWIW we haven't experiencing any of this with jitsi meet.

**Comment by renandincer** [comment]
We don't do anything different for Chrome vs react-native-webrtc so I'd expect them to work the same. Looking at the diff above, `encoderImplementation` stands out across the two tests.

**Comment by saghul** [comment]
If anyone can give the latest release (124.0.0) a try and report back, that would be appreciated!

**Comment by skirsten** [comment]
I just did a quick test with version `124.0.1` and unfortunately the issue persists.

If I can find the time I will do some more investigation with different sources to track down if the issue is on the encoding or decoding side and/or is related to the codec etc.

**Comment by saghul** [comment]
Thanks for keeping at it, I haven't found the time to dig into it yet myself.

I do suspect a combination with something CF does because nobody else has reported this.

**Comment by skirsten** [comment]
Hey, just noticed this issue at the bottom of my notifications and just wanted to give an update.

I am pretty sure the issue only appears when playing back the same stream that is being sent from the same device (like in the example of this repo). If the receiving and sending stream is different (like in most real apps) I did not encounter any problems.

I then tried to trace it down to something specific. My guess was that maybe some algorithm was seeing the same ID on the receiving and sending side causing the feedback loop. Or that the resolution change when it changes quality causes a slight lag that could maybe impact the encoder side causing a feedback loop again.

Maybe when using the local connection (like in the examples in this repo) instead of via CF the latency was so low that this feedback loop did not kick in or the feedback loop is indeed on CF side when receiving and sending the same stream to the same client.

Honestly I have no idea...

This was a few months ago and unfortunately I could not find a specific issue and have not worked on the project since then.

So I will close this issue as probably it has nothing to do with Cloudflare Calls specifically and should not occur in real-world apps. If there is need to reopen, feel free to do so. Thanks!
